### PR TITLE
Refactor of Success and Failure Blocks, Added Errors with Debugging Levels

### PR DESCRIPTION
**Pull request by need2edit**
- added debugging level to the global API with a default of none to not
  affect existing code
- added print statements for account errors
- added print statements for empty projects
- added print statements for projects with a bad hashed id
- fed error from Alamofire in to print statements where appropriate

### CODE REVIEW

File: Pod/Classes/public/error/Error.swift [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    switch debuggingLevel {
+        
+    case .None:
+        break
+    case .Friendly:
+        
+        // We print the message and fallthrough to allow .Annoying case to add to things.
+        guard let message = message else { break }
+        print(message)
+        fallthrough
```
Comment (spinosa) [inline Pod/Classes/public/error/Error.swift]: This `fallthrough` affects the `.Friendly` case.  So, if debugging is `.Friendly`, it falls though and executes the `.Annoying` case's code as well.  From a cursory read, it seems this wasn't the intention (ie. annoying should print everything, friendly should print less).

File: Pod/Classes/public/error/Error.swift [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    case None
+    
+    /// Print the high level activity and errors.
+    case Friendly
+    
+    /// Print all the things.
+    case Annoying
+}
+
+internal func printErrorForDebuggingLevel(debuggingLevel debuggingLevel: DebuggingLevel, message: String? = nil, error: NSError? = nil) {
```
Comment (spinosa) [inline Pod/Classes/public/error/Error.swift]: I'd like to follow the Swift 3 [guidelines for APIs](https://swift.org/documentation/api-design-guidelines/)

In this case, let's drop `DebuggingLevel` from the function name as it will already be included as the name of the first parameter.

File: Pod/Classes/public/error/Error.swift [hunk 3]
```diff
@@ -0,0 +1,150 @@
+//
+//  Error.swift
+//  Pods
+//
+//  Created by Jake Young on 6/14/16.
+//
```
Comment (spinosa) [inline Pod/Classes/public/error/Error.swift]: Please update header block to include Wistia copyright and WistiaKit project name (not Pods)

File: README.md [hunk 4]
```diff
@@ -111,7 +111,7 @@ For those of you who want total control, you want the `WistiaPlayer`.  To see th
...
+* `requireHLS` - Apple has [specific requirements](https://developer.apple.com/app-store/review/guidelines/#media-content) for playing video within apps.  It boils down to this: if you want to play video over 10 minutes in length over the celluar network (ie. you don't force wifi), then you must use [HLS](https://developer.apple.com/streaming/).  And Wistia fully supports and has thoroughly tested our HLS implementation with Apple.   We recommend leaving this to `true` which is the default. HLS must be enabled on your Wistia account to take advantage of this feature.  If you don't have it enabled, set this to `false`.  Not sure if HLS is enabled on your account? Contact us at [support@wistia.com](mailto:support@wistia.com?subject='Is HLS enabled on my Wistia Account?')!
```
Comment (spinosa) [inline README.md]: 👍

File: Pod/Classes/public/WistiaAPI.swift [hunk 5]
```diff
@@ -41,8 +42,9 @@ public class WistiaAPI {
 
      - Parameter apiToken: The API Token that will be used to access the [Wistia Data API](http://wistia.com/doc/data-api).
      */
-    public init(apiToken:String) {
+    public init(apiToken:String, debuggingLevel: DebuggingLevel = .None) {
```
Comment (spinosa) [inline Pod/Classes/public/WistiaAPI.swift]: Please update function documentation for `debuggingLevel`

File: Pod/Classes/public/WistiaAPI.swift [hunk 6]
```diff
@@ -105,11 +107,26 @@ extension WistiaAPI {
...
+                            account = ModelBuilder.accountFromHash(JSON) {
+                            completionHander(account: account)
+                        }
+                    
+                    case .Failure(let error):
+                        
+                        // Let the user know there is an issue.
+                        printErrorForDebuggingLevel(
+                            debuggingLevel: self.debuggingLevel,
+                            message: WistiaAccount.Error.InvalidAPIPassword.description,
```
Comment (spinosa) [inline Pod/Classes/public/WistiaAPI.swift]: Request could have failed for other reasons, no?  Timeout, 500, etc.?

File: Pod/Classes/public/WistiaAPI.swift [hunk 7]
```diff
@@ -105,11 +107,26 @@ extension WistiaAPI {
...
 
-                if let JSON = response.result.value as? [String: AnyObject],
-                    account = ModelBuilder.accountFromHash(JSON) {
-                    completionHander(account: account)
-                } else {
-                    completionHander(account: nil)
+                // 'Swift-like' way to handle Alamofire responses
+                switch response.result {
+                    
+                    // Success! Let's parse that JSON.
```
Comment (spinosa) [inline Pod/Classes/public/WistiaAPI.swift]: When the code is sufficiently self-documenting, we don't need additional comments.  Indeed, it's safer not to have them as they can get stale.

I would drop this one and some of the others added in this class.

File: Pod/Classes/public/WistiaAPI.swift [hunk 8]
```diff
@@ -139,19 +156,34 @@ extension WistiaAPI {
...
+                
+                // 'Swift-like' way to handle Alamofire responses
+                switch response.result {
+                    
+                    // Success! Let's parse that JSON.
+                    case .Success(let value):
+                        
+                        if let JSON = value as? [[String: AnyObject]] {
+                            
+                            // We map each JSON to WistiaProject objects, and filter out any nil values
```
Comment (spinosa) [inline Pod/Classes/public/WistiaAPI.swift]: example of a good comment since code is terse

File: Pod/Classes/public/WistiaAPI.swift [hunk 9]
```diff
@@ -139,19 +156,34 @@ extension WistiaAPI {
...
+                            let projects = JSON.flatMap { ModelBuilder.projectFromHash($0) }
+                            
+                            // Let the user know if a project is empty.  This isnt a bug per se, but if the user is expecting something and there are no projects, it a good opportunity to inform.
```
Comment (spinosa) [inline Pod/Classes/public/WistiaAPI.swift]: great comment

File: Pod/Classes/public/error/Error.swift [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    }
+    
+}
+
+
+/// The `Error` struct provides a convenience for creating custom WistiaKit NSErrors and Swift ErrorTypes.
+public struct Error {
+    
+    /// The domain used for creating all Wistia errors.
+    public static let Domain = "com.wistia.error"
```
Comment (spinosa) [inline Pod/Classes/public/error/Error.swift]: Can you explain the reasoning for this?

[Don't read that as a snarky way to say "drop this."  I'd like to learn / this is honest curiosity.]

**Comment by spinosa** [comment]
Hey @need2edit,  Love this!

A few little comments inline, but there are two higher level things:

1) Travis CI build failed.

2)  I'm trying to get insight into `Error`, `ErrorTypeWithDomain`, and `ErrorProvider`.  I see how they're being used, but I'm not grok'ing why.  Could you help me understand the design?

File: Pod/Classes/public/error/Error.swift (hunk 10 above)
Comment (need2edit) [inline Pod/Classes/public/error/Error.swift]: @spinosa thanks for inquiring, the Error struct here might make a whole lot more sense if it had convenience methods to generate NSErrors from these Wistia Errors.

Apple has done a lot of work to automatically bridge ErrorType and NSError, so I might be convincing myself this isn't necessary within the current intended focus for WistiaKit.  With ErrorType paired with enums, its a better design to pass messages etc. but while its "bridged" to NSError, you lose any associated information with it. 

In other words, .MyError("All the things failed.") in Swift loses its associated message when sent over to NSError... when really you want to convert the message and store it in the "userInfo" dictionary for NSError.

Basically, domain, code, etc, would be used to generate NSError objects for situations where NSErrors are expected... so maybe its a question of "how much Objective-C interoperability does WistiaKit need to have?"

Comment (spinosa) [inline Pod/Classes/public/error/Error.swift]: Apple is pushing Swift hard.  It's the future, if not the present.  I'm leaning heavily in favor of following that pattern unless/until there is a very compelling reason to increase ObjC interop.

Dropping NSError (and its related additions) makes this simpler and clearer.

Comment (need2edit) [inline Pod/Classes/public/error/Error.swift]: Agreed (especially after a week of WWDC updates and best practice), removing this and revising.

File: Pod/Classes/public/error/Error.swift (hunk 2 above)
Comment (need2edit) [inline Pod/Classes/public/error/Error.swift]: @spinosa after revisiting this the debugging levels are nice to have, but the real need is to print our what's going on or not, so resubmitting with debug boolean, and applying Swift 3 guidelines.

**Comment by spinosa** [comment]
Hey @need2edit!  WistiaKit v 0.17.1 now includes error handling.  Your feedback would be appreciated.

**Comment by need2edit** [comment]
@spinosa thanks for the call out! Taking a look over the next few days! Great work!
